src/pydoc_markdown/contrib/renderers/smartresolver.py

```diff
@@ -0,0 +1,381 @@
+                    return None
+        
+
+
+        for _import in _imports_all:
+
+            if _import.Process():
+
+                if _import.module == previousModule:
+                    # recursive, won't be any good
+                    return None
+
+                result = self.FindRefFromImport(_import.module, refNames, typeFilter, module)
+                if result: return result
+            else:
+                return None
+        
+        return None
+
+    def FindPageByName(self, pages : Pages[Page], target : str) -> Pages[Page]:
+
+        for page in pages:
+
+            if page.contents:
+                for content in page.contents:
+                    
+                    if content == target or content == "*":
+                        return [page]
+
+                    elif content[-2:] == ".*":
+                        if target.startswith(content[:-2]):
+                            return [page]
+                            
+                    elif content == "*":
+                        return
+
+            if len(page.children) > 0:
+                result = self.FindPageByName(page.children, target)
+                if result: return [page] + result
+        return None
+        
+    def FindTargetPageURL(self, obj: docspec.ApiObject, target : docspec.ApiObject) -> t.Optional[str]:
+        
+        obj_id = self.GetObjectID(obj)
+        target_id = self.GetObjectID(target)
+
+        this_page = self.FindPageByName(self.pages, obj_id)
+        target_page = self.FindPageByName(self.pages, target_id)
+
+        if target_page == None or this_page == None: return None
+
+        for parent_index in range(len(this_page)):
+                if parent_index >= len(target_page): break
+                if this_page[parent_index] != target_page[parent_index]:
+                        parent_index -= 1
+                        break
+        
+        parent_index += 1
+        
+        relative_path = "../" * (len(this_page) - parent_index)
+        url = relative_path + "/".join(page.title.replace(" ", "-").lower() for page in target_page[parent_index:]) + "#" + target_id
+        
+        return url
+
+    def resolve_ref(self,
+                    obj        : docspec.ApiObject,
+                    reference  : str,
+                    typeFilter : t.List[t.Type[docspec.ApiObject]] = []
+        ) -> t.Optional[t.Tuple[str, docspec.ApiObject]]:
+        
+        refNames = reference.split(".")
+        targetURL = None
+
+        # look for the reference in the local module
+        if self.renderer.crossref_local:
+            result = self.FindRefLocal(obj, refNames, typeFilter)
+            if result: targetURL = self.FindTargetPageURL(obj, result)
+        
+        if targetURL: return targetURL
+
+        # look for the reference in all of the imported modules
+        if self.renderer.crossref_import:
+            result = self.FindRefFromImport(obj, refNames, typeFilter)
```

Comment:
I haven't played a lot with it yet, but so far I couldn't get references a cross pages to work and I think that is because I tried to use them for classes that are also in different modules, and the docspec tree filtering that is hapenning before the renderer receives the modules causes the `docspec.Indirection` objects to be stripped from the tree and thus this function won't be able to find them at all.